route_following_plugin/src/route_following_plugin.cpp

```diff
@@ -41,190 +41,238 @@ namespace route_following_plugin
+            double target_speed_in_lanelet = findSpeedLimit(route_shortest_path[i]);
+            double start_dist = wm_->routeTrackPos(route_shortest_path[i].centerline2d().front()).downtrack;
+            double end_dist = wm_->routeTrackPos(route_shortest_path[i].centerline2d().back()).downtrack; 
 
-        //Update current status based on prior plan
-        if(req.prior_plan.maneuvers.size()!=0){
-            time_progress = req.prior_plan.planning_completion_time;
-            int end_lanelet =0;
-            updateCurrentStatus(req.prior_plan.maneuvers.back(),speed_progress,current_progress,end_lanelet);
-            last_lanelet_index = findLaneletIndexFromPath(end_lanelet,shortest_path);
-        }
+            end_dist = std::min(end_dist, route_length);
 
-        bool approaching_route_end = false;
-        double time_req_to_stop,stopping_dist;
-        time_req_to_stop = sqrt(2*findSpeedLimit(shortest_path.back())/jerk_); 
-        stopping_dist = findSpeedLimit(shortest_path.back())*time_req_to_stop - (0.167 * jerk_ * pow(time_req_to_stop,3));
-        
-        if(route_length - current_progress <= stopping_dist){
-            approaching_route_end = true;
-        }
-        ROS_DEBUG_STREAM("Starting Loop");
-        ROS_DEBUG_STREAM("Time Required To Stop: " << time_req_to_stop << " stopping_dist: " << stopping_dist<<" Speed limit:"<<findSpeedLimit(shortest_path.back())<< " Jerk:"<<jerk_);
-        ROS_DEBUG_STREAM("total_maneuver_length: " << total_maneuver_length << " route_length: " << route_length);
-        while(current_progress < total_maneuver_length && !approaching_route_end)
-        {
-            ROS_DEBUG_STREAM("Lanlet: " << shortest_path[last_lanelet_index].id());
-            ROS_DEBUG_STREAM("current_progress: "<< current_progress);
-            ROS_DEBUG_STREAM("speed_progress: " << speed_progress);
-            ROS_DEBUG_STREAM("target_speed: " << target_speed);
-            ROS_DEBUG_STREAM("time_progress: " << time_progress.toSec());
-            auto p = shortest_path[last_lanelet_index].centerline2d().back();
-            double end_dist = wm_->routeTrackPos(shortest_path[last_lanelet_index].centerline2d().back()).downtrack;
-            end_dist = std::min(end_dist, total_maneuver_length);
-            ROS_DEBUG_STREAM("end_dist: " << end_dist);
-            double dist_diff = end_dist - current_progress;
-            ROS_DEBUG_STREAM("dist_diff: " << dist_diff);
-            if(route_length - end_dist <= stopping_dist){
-                end_dist = route_length - stopping_dist; 
-                dist_diff = end_dist - current_progress;
-                approaching_route_end = true;
+            if(isLaneChangeNeeded(following_lanelets, route_shortest_path[i + 1].id())){
+                maneuvers.push_back(composeLaneChangeManeuverMessage(start_dist, end_dist, prev_mvr_end_speed , target_speed_in_lanelet, route_shortest_path[i].id(), route_shortest_path[i+1].id(), ros::Time::now()));
```

Comment:
I'm not sure that ros::Time::now() is accurate, considering this could be the second or third or so-on maneuver in the maneuver plan, for which the start time would need to be ~= to the end time of the maneuver before it.

Reply:
This starting time is updated in plan_maneuver_cb before the maneuvers are added to response